### PR TITLE
stdenv: Fix overriding + `overrideAttrs`

### PR DESCRIPTION
**Pull request by Ericson2314**
###### Motivation for this change

The old stdenv adapters were subtly wrong in two ways:

 - `overrideAttrs` leaked the original, unoverridden `mkDerivation`.

 - `stdenv.override` would throw away any manually-set `mkDerivation`
   from a stdenv reverting to the original.

Now, `mkDerivation` is controlled (nearly directly) via an argument, and
always correctly closes over the final ("self") stdenv. This means the
adapters can work entirely via `.override` without any manual `stdenv //
...`, and both those issues are fixed.

Note hashes are changed, because stdenvs no previously overridden like
`stdenvNoCC` and `crossLibcStdenv` now are. I had to add some
`dontDisableStatic = true` accordingly. The flip side however is that
since the overrides compose, we no longer need to override anything but
the default `stdenv` from which all the others are created.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/stdenv/generic/default.nix [hunk 1]
```diff
@@ -48,6 +48,10 @@ let lib = import ../../../lib; in lib.makeOverridable (
 
 , # The platform which build tools (especially compilers) build for in this stage,
   targetPlatform
+
+, # The implementation of `mkDerivation`, parameterized with the final stdenv so we can tie the knot.
+  # This is convient to have as a parameter so the stdenv "adapters" work better
```
Comment (Mindavi) [inline pkgs/stdenv/generic/default.nix:53]: Typo: convenient

File: pkgs/stdenv/adapters.nix [hunk 2]
```diff
@@ -2,7 +2,31 @@
...
 
-pkgs:
+{ lib, pkgs, config }:
+
+let
+  # N.B. Keep in sync with default arg for stdenv/generic.
+  defaultMkDerivationFromStdenv = import ./generic/make-derivation.nix { inherit lib config; };
+
+  # Low level function to help with overriding `mkDerivationFromStdenv`. One
+  # gives it the old stdenv arguments and a "continuation" funciont, and
```
Comment (Pamplemousse) [inline pkgs/stdenv/adapters.nix:12]: Small typo: `funciont` --> `function`

File: pkgs/stdenv/adapters.nix [hunk 3]
```diff
@@ -2,7 +2,31 @@
...
+  # underneath the final stdenv argument it yields to the continuation to do
+  # whatever it wants with old `mkDerivation` (old `mkDerivationFromStdenv`
+  # applied to the *new, final* stdenv) provided for convience.
```
Comment (Pamplemousse) [inline pkgs/stdenv/adapters.nix:15]: Small typo: `convience` --> `convenience`

File: pkgs/stdenv/generic/default.nix [hunk 4]
```diff
@@ -48,6 +48,10 @@ let lib = import ../../../lib; in lib.makeOverridable (
...
+  mkDerivationFromStdenv ? import ./make-derivation.nix { inherit lib config; }
```
Comment (Pamplemousse) [inline pkgs/stdenv/generic/default.nix:54]: Is it necessary to give a default value given that `mkDerivationFromStdenv-super` defaults to `defaultMkDerivationFromStdenv` in `adapters.nix`?

File: pkgs/stdenv/generic/make-derivation.nix [hunk 5]
```diff
@@ -7,7 +9,7 @@ let
     # to build it. This is a bit confusing for cross compilation.
     inherit (stdenv) hostPlatform;
   };
-in rec {
+in
   # `mkDerivation` wraps the builtin `derivation` function to
```
Comment (Pamplemousse) [inline pkgs/stdenv/generic/make-derivation.nix:13]: This comment / documentation placement is confusing now...

File: pkgs/stdenv/adapters.nix (hunk 3 above)
Comment (Mindavi) [inline pkgs/stdenv/adapters.nix:15]: Typo: convenience

File: pkgs/stdenv/generic/default.nix (hunk 4 above)
Comment (Ericson2314) [reply pkgs/stdenv/generic/default.nix:54]: The non-adapted stdenv needs this. I sort of see it as the normal one and the other as the unfortunate cludge too.

**Review by Mindavi — COMMENTED** [review]
I don't know much about the stdenv, but added some nits about typos :)

Edit: pamplemousse was faster

File: pkgs/stdenv/adapters.nix [hunk 6]
```diff
@@ -2,7 +2,31 @@
...
+  withOldMkDerivation = stdenvSuperArgs: k: stdenvSelf: let
```
Comment (Pamplemousse) [inline pkgs/stdenv/adapters.nix:16]: First argument name is confusing me a little.
Could it just be `stdenvSuper`?

File: pkgs/stdenv/generic/make-derivation.nix (hunk 5 above)
Comment (Ericson2314) [reply pkgs/stdenv/generic/make-derivation.nix:13]: I was thinking of fixing that in a separate commit to help with git blame, but I should make it part of this PR yes.

File: pkgs/stdenv/adapters.nix (hunk 6 above)
Comment (Ericson2314) [reply pkgs/stdenv/adapters.nix:16]: It's not the stdenv itself, but the arguments passed to `generic/default.nix`.

Comment (Pamplemousse) [reply pkgs/stdenv/adapters.nix:16]: Ah yes!

File: pkgs/stdenv/adapters.nix [hunk 7]
```diff
@@ -31,52 +55,52 @@ rec {
...
   # shared libraries.
-  makeStaticLibraries = stdenv: stdenv //
-    { mkDerivation = args: stdenv.mkDerivation (args // {
+  makeStaticLibraries = stdenv:
+    stdenv.override (old: {
+      mkDerivationFromStdenv = extendMkDerivationArgs old (args: {
         dontDisableStatic = true;
-      } // pkgs.lib.optionalAttrs (!(args.dontAddStaticConfigureFlags or false)) {
+      } // lib.optionalAttrs (!(args.dontAddStaticConfigureFlags or false)) {
         configureFlags = (args.configureFlags or []) ++ [
```
Comment (Pamplemousse) [inline pkgs/stdenv/adapters.nix:84]: Just for reference, following a quick chat @Ericson2314 and I had:
For packages that unconditionally add `--enable-shared` in their `configureFlags` (for example: [`musl`](https://github.com/NixOS/nixpkgs/blob/170ea681e5807e81788e924b8259d2a08be4b48a/pkgs/os-specific/linux/musl/default.nix#L74-L80)), this will effectively add both `--enable-shared --disable-shared` to the compilation command...

**Comment by symphorien** [comment]
This break musl somehow:
```
./tools/install.sh -D obj/ld.musl-clang /nix/store/v27r72r03h873j895qg9phq4j6175msq-musl-static-x86_64-unknown-linux-musl-1.2.2/bin/ld.musl-clang
Moving /nix/store/v27r72r03h873j895qg9phq4j6175msq-musl-static-x86_64-unknown-linux-musl-1.2.2/bin/musl-gcc to /nix/store/iphgzmq6c5fi5pr84lh9sbzmhz8aikwa-musl-static-x86_64-unknown-linux-musl-1.2.2-dev/bin/musl-gcc
rmdir: failed to remove '/nix/store/v27r72r03h873j895qg9phq4j6175msq-musl-static-x86_64-unknown-linux-musl-1.2.2/bin': Directory not empty
Moving /nix/store/v27r72r03h873j895qg9phq4j6175msq-musl-static-x86_64-unknown-linux-musl-1.2.2/bin/musl-clang to /nix/store/iphgzmq6c5fi5pr84lh9sbzmhz8aikwa-musl-static-x86_64-unknown-linux-musl-1.2.2-dev/bin/musl-clang
rmdir: failed to remove '/nix/store/v27r72r03h873j895qg9phq4j6175msq-musl-static-x86_64-unknown-linux-musl-1.2.2/bin': Directory not empty
Moving /nix/store/v27r72r03h873j895qg9phq4j6175msq-musl-static-x86_64-unknown-linux-musl-1.2.2/bin/ld.musl-clang to /nix/store/iphgzmq6c5fi5pr84lh9sbzmhz8aikwa-musl-static-x86_64-unknown-linux-musl-1.2.2-dev/bin/ld.musl-clang
rmdir: failed to remove '/nix/store/v27r72r03h873j895qg9phq4j6175msq-musl-static-x86_64-unknown-linux-musl-1.2.2/bin': Directory not empty
Moving /nix/store/v27r72r03h873j895qg9phq4j6175msq-musl-static-x86_64-unknown-linux-musl-1.2.2/lib/musl-gcc.specs to /nix/store/iphgzmq6c5fi5pr84lh9sbzmhz8aikwa-musl-static-x86_64-unknown-linux-musl-1.2.2-dev/lib/musl-gcc.specs
rmdir: failed to remove '/nix/store/v27r72r03h873j895qg9phq4j6175msq-musl-static-x86_64-unknown-linux-musl-1.2.2/lib': Directory not empty
/nix/store/ak7vllbyx80qz6ifhjvipxgai7axsay9-x86_64-unknown-linux-musl-binutils-2.35.1/bin/x86_64-unknown-linux-musl-ld:INSTALL: file format not recognized; treating as linker script
/nix/store/ak7vllbyx80qz6ifhjvipxgai7axsay9-x86_64-unknown-linux-musl-binutils-2.35.1/bin/x86_64-unknown-linux-musl-ld:INSTALL:2: syntax error
collect2: error: ld returned 1 exit status
builder for '/nix/store/zwppa911qs55y0pv3xs1iyn7rqfa1ydm-musl-static-x86_64-unknown-linux-musl-1.2.2.drv' failed with exit code 1
```

File: pkgs/stdenv/adapters.nix (hunk 7 above)
Comment (symphorien) [reply pkgs/stdenv/adapters.nix:84]: yes this regression has been there since the introduction of dontAddStaticConfigureFlags

Comment (symphorien) [reply pkgs/stdenv/adapters.nix:84]: I'm fixing a number of them in #133008

**Comment by Pamplemousse** [comment]
@symphorien I think `pkgsStatic.musl` was broken before that.
I tried fixing it yesterday, but there are many things seemingly going wrong in the `postInstall` script...

cc @dtzWill @thoughtpolice (maintainers of `musl`)

**Comment by symphorien** [comment]
well I rebased #133008 on this and got new failures due to that, so there must be a regression somewhere...

**Comment by symphorien** [comment]
On the parent commit of this pr, pkgsStatic.hello build, but not on this commit. The failure involves musl but maybe it's a red herring, since as you pointed out pkgsStatic.musl has not been building for a long time.

**Comment by Pamplemousse** [comment]
@symphorien My guess is that this PR has just made the failure of `pkgsStatic.musl` apparent (i.e. happen when building other `pkgsStatic` packages - as if `musl` was used before, and now `pkgsStatic.musl` is).

**Comment by Ericson2314** [comment]
OK fixed the issue. The derivation to check is `pkgsStatic.muslCross` (which is separate for stupid reasons; hope to finally get rid of that with #132343). Comparing it with `nix-diff` we see that since `stdenvNoCC` and `crossLibcStdenv` now pick up the overrides from `stdenv`, which means that Musl no longer was building the dynamic linker, which meant that the on-the-fly compilation of `iconv` was failing.

No, I am not sure why we want the dynamic build of Musl, but I didn't want to make a big behavioral change like that, so I just added `dontAddStaticConfigureFlags = true` instead.

**Comment by r-burns** [comment]
I think this somehow removed `pkgsStatic.__bootPackages`:

```
> nix-build . -A pkgsStatic.stdenv.__bootPackages 
error: attribute '__bootPackages' in selection path 'pkgsStatic.stdenv.__bootPackages' not found
```

**Comment by r-burns** [comment]
Oh, I think this is happening because `__bootPackages` and `__hatPackages` are added to each stdenv stage using `//`. 

https://github.com/NixOS/nixpkgs/blob/abeef13d12fbc2eecc75c1138bdcc0b9ad22b32d/pkgs/stdenv/booter.nix#L92-L96

So if stdenv is overridden, these extra attributes are discarded.

**Comment by Ericson2314** [comment]
Yup that's why. I feel like the cure might be worse than the disease. In any event, I hope we might finally get away from that and the way bootstrapping is done today once https://github.com/NixOS/nixpkgs/pull/132343 is done.

**Comment by teto** [comment]
I can confirm this ^ . Breaks `enableDebugging` too (since it calls `keepDebugInfo`).

**Comment by trofi** [comment]
Should `enableDebugging` still work? Comment says it should:

```
  # intended to be used like nix-build -E 'with import <nixpkgs> {}; enableDebugging fooPackage'
  enableDebugging = pkg: pkg.override { stdenv = stdenvAdapters.keepDebugInfo pkg.stdenv; };
```

Filed https://github.com/NixOS/nixpkgs/issues/136756

**Comment by roberth** [comment]
We should seriously consider simplifying the problem away. https://github.com/NixOS/nixpkgs/issues/136903#issuecomment-915883835
I have no confidence that we can use a redundant layer of overriding correctly.
tl;dr: we can remove the fixpoint we don't need by improving the one we do need, making`overrideAttrs` more powerful.

**Comment by Ericson2314** [comment]
@bluescreen303 See #136756

File: pkgs/stdenv/adapters.nix [hunk 8]
```diff
@@ -31,52 +55,52 @@ rec {
...
         NIX_CFLAGS_LINK = toString (args.NIX_CFLAGS_LINK or "") + " -static";
-      } // pkgs.lib.optionalAttrs (!(args.dontAddStaticConfigureFlags or false)) {
+      } // lib.optionalAttrs (!(args.dontAddStaticConfigureFlags or false)) {
         configureFlags = (args.configureFlags or []) ++ [
             "--disable-shared" # brrr...
           ];
-      });
-    };
+      }));
+    } // lib.optionalAttrs (stdenv0.hostPlatform.libc == "libc") {
```
Comment (amjoseph-nixpkgs) [inline pkgs/stdenv/adapters.nix:70]: @Ericson2314 did you mean `"glibc"` here instead of `"libc"`?

@picostove reports that changing it fixes #244232.

Comment (Artturin) [reply pkgs/stdenv/adapters.nix:70]: He did mean glibc https://github.com/NixOS/nixpkgs/pull/134923#discussion_r693084820

> since that one had the typo, neither is doing anything.

Comment (Artturin) [reply pkgs/stdenv/adapters.nix:70]: https://github.com/NixOS/nixpkgs/pull/244403
